### PR TITLE
rbd: provide option to disable setting metadata on rbd images

### LINKED ISSUE #3009: Provide option to disable setting metadata on rbd images
As we added support to set the metadata on the rbd images created for the PVC and volume snapshot in https://github.com/ceph/ceph-csi/pull/2898 by default metadata is set on all the images. As we have seen we are hitting https://github.com/ceph/ceph-csi/issues/2327 a lot of times with #2898. with this, we start to leave a lot of stale images. Currently, we rely on https://github.com/ceph/ceph-csi/blob/4100ad24061670c44fe94f1a85de8dcd34d237c9/deploy/rbd/kubernetes/csi-rbdplugin-provisioner.yaml#L61 to decide to set the metadata or not, we cannot set this option to false to disable setting metadata because we use this for encryption also. This is to request to provide an option to disable setting the image metadata when starting cephcsi.

### PR DESCRIPTION
**Pull request by pkalever**
# Describe what this PR does #

As we added support to set the metadata on the rbd images created for the PVC and volume snapshot, by default metadata is set on all the images.
                                                                          
As we have seen we are hitting issue#2327 a lot of times with this, we start to leave a lot of stale images. Currently, we rely on            
`--extra-create-metadata=true` to decide to set the metadata or not, we cannot set this option to false to disable setting metadata because we use this for encryption too.                                              
                                                                          
This changes is to provide an option to disable setting the image metadata when starting cephcsi.                                           
                                                                          

## Related issues ##

Fixes: #3009                                                              

Signed-off-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>       

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

### CODE REVIEW

File: charts/ceph-csi-rbd/values.yaml
Comment (Madhu-1) [inline charts/ceph-csi-rbd/values.yaml:214]: disable it by default until the ceph issue is fixed?

Comment (Madhu-1) [inline charts/ceph-csi-rbd/values.yaml:214]: Need documentation update of helm parameters in Readme.md

File: cmd/cephcsi.go
Comment (Madhu-1) [inline cmd/cephcsi.go]: I think this won't exceed the line length limit. it can be a single statement like others,

File: internal/rbd/controllerserver.go
Comment (Madhu-1) [inline internal/rbd/controllerserver.go]: can we make `SetMetadata` part of the rbdImage struct, if we do so we just need if check inside the  `setSnapshotMetadata` and `setVolumeMetadata`, and changes will be minimal and easy to remove this option later also? WDYT?

**Review by Madhu-1 — DISMISSED** [review]
few nits and questions.

File: internal/rbd/controllerserver.go
Comment (pkalever) [reply internal/rbd/controllerserver.go]: Sounds good to me.

File: charts/ceph-csi-rbd/values.yaml
Comment (pkalever) [reply charts/ceph-csi-rbd/values.yaml:214]: We should comment out the test cases then?

**Comment by nixpanic** [comment]
Just wondering, why a commandline option, and not a parameter in the StorageClass?

File: internal/controller/controller.go
Comment (nixpanic) [inline internal/controller/controller.go]: comments are ideally above the attribute

File: internal/rbd/rbd_util.go
Comment (nixpanic) [inline internal/rbd/rbd_util.go]: Do the reverse check, and return if setting the metadata is not needed. It is like testing for an error, and returning before executing further operations.

That reduces the indention for the next code block. It is the style we try to use throughout the project.

**Review by nixpanic — CHANGES_REQUESTED** [review]
I'm not sure what the plan was with this PR. The option is deployment wide, whereas it could also be per StorageClass.

If I'm not mistaken, this is only required because there is a Ceph bug that gets hit regularly when the metadata is set on the RBD images? Link to the Ceph bug, and find out when a release includes a fix?

**Comment by pkalever** [comment]
Rebased, solving many merge conflicts! I will test this manually and update the behavior shortly.

File: charts/ceph-csi-rbd/values.yaml
Comment (pkalever) [reply charts/ceph-csi-rbd/values.yaml:214]: Ah! Noticed this now :-)
Still need to address it.

Comment (pkalever) [reply charts/ceph-csi-rbd/values.yaml:214]: Done now.

**Comment by pkalever** [comment]
@Madhu-1 I got the same struct padding error now, see https://github.com/ceph/ceph-csi/runs/6916598994?check_suite_focus=true

**Comment by pkalever** [comment]
> Rebased, solving many merge conflicts! I will test this manually and update the behavior shortly.

Testing is done, and the changes are working as expected.

**Comment by pkalever** [comment]
@mergifyio rebase

**Comment by pkalever** [comment]
@Mergifyio rebase

**Review by Rakshith-R — DISMISSED** [review]

**Review by nixpanic — DISMISSED** [review]

**Comment by nixpanic** [comment]
@mergifyio rebase

**Comment by pkalever** [comment]
[Deployment failure
](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.21/detail/mini-e2e-helm_k8s-1.21/4344/pipeline)/retest ci/centos/mini-e2e-helm/k8s-1.21

**Comment by pkalever** [comment]
/retest ci/centos/mini-e2e-helm/k8s-1.23
```
Jun 21 10:32:16.449: FAIL: backend images not matching kubernetes resource count,image count 22 kubernetes resource count 21
```
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e-helm_k8s-1.23/runs/2188/nodes/107/log/?start=0)

File: internal/rbd/rbd_util.go
Comment (Madhu-1) [inline internal/rbd/rbd_util.go:2062]: @pkalever same check is also required in unsetMetadata to skip RemoveMetadata call or else we might end up calling removemetadata and discard not found erro?

Comment (pkalever) [reply internal/rbd/rbd_util.go:2062]: I had intentionally not mentioned that in the RemoveMetadata, because if the user had enabled setmetadata at some point in time and disabled it later, the calls coming to RemoveMetadata should be handled and metadata should be removed on previous images (in case metadata exists on them)?

Comment (Madhu-1) [reply internal/rbd/rbd_util.go:2062]: we can document the expectations WDYT?. adding check-in unsetMetadata helps in avoiding 3 calls RemoveMetadat call.

Comment (pkalever) [reply internal/rbd/rbd_util.go:2062]: IMHO, it is our csi responsibility to clean the metadata that it has added. Moreover, in case metadata is absent on the image, it will skip the rest of the operation to remove metadata. But that all said, I would like to leave the final decision to you!

Comment (Madhu-1) [reply internal/rbd/rbd_util.go:2062]: @nixpanic any suggestion?

Comment (nixpanic) [reply internal/rbd/rbd_util.go:2062]: re-considered this approach, and I think it is cleanest if the caller decides if the metadata needs to be set. The `EnableMetadata` is not really an option for the `rbdImage`, but rather the Controller or Node-Plugin.

If the option is changed after volumes have been created with metadata, keeping stale metadata is acceptable, it should just not be used.

Comment (pkalever) [reply internal/rbd/rbd_util.go:2062]: Cool, I will consider adding `if !rv.EnableMetadata{}` defence check in removeMetadata() too, Thanks guys!

**Comment by pkalever** [comment]
@nixpanic @Madhu-1 @Rakshith-R please replace your acks!

**Diff**
```
[0] pkalever 😎 ceph-csi✨ git diff     
diff --git a/internal/rbd/rbd_util.go b/internal/rbd/rbd_util.go
index e8e1f40a9..59c26f187 100644          
--- a/internal/rbd/rbd_util.go              
+++ b/internal/rbd/rbd_util.go
@@ -2087,6 +2087,10 @@ func (rv *rbdVolume) setAllMetadata(parameters map[string]string) error {
                                       
 // unsetAllMetadata unset all the metadata from arg keys on RBD image.
 func (rv *rbdVolume) unsetAllMetadata(keys []string) error {
+       if !rv.EnableMetadata {                       
+               return nil                                                                                             
+       }                              
+                                                                                                                      
        for _, key := range keys {
                err := rv.RemoveMetadata(key)
                // TODO: replace string comparison with errno.
```

**Review by Madhu-1 — DISMISSED** [review]

**Review by Rakshith-R — DISMISSED** [review]

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Comment by pkalever** [comment]
@Madhu-1 @nixpanic @ceph/ceph-csi-contributors 
Seeing this, not sure why!
```
[1mSTEP[0m: create a PVC and check PVC/PV metadata on RBD image
[...]
Jun 24 04:51:27.094: INFO: stdErr occurred: failed to get metadata csi.storage.k8s.io/pvc/name of image : (2) No such file or directory
rbd: getting metadata failed: (2) No such file or directory

Jun 24 04:51:27.094: FAIL: failed to get PVC name --pool=replicapool/csi-vol-ea2d0eb1-f370-11ec-8161-0ed2bb196630 csi.storage.k8s.io/pvc/name: err=command terminated with exit code 2 stdErr="failed to get metadata csi.storage.k8s.io/pvc/name of image : (2) No such file or directory\nrbd: getting metadata failed: (2) No such file or directory\n"
[AfterEach] RBD
```
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/rest/organizations/jenkins/pipelines/mini-e2e_k8s-1.23/runs/2081/nodes/89/log/?start=0)

**Comment by pkalever** [comment]
@ceph/ceph-csi-contributors Here is the issue to track the e2e changes https://github.com/ceph/ceph-csi/issues/3220

File: charts/ceph-csi-rbd/README.md
Comment (Madhu-1) [inline charts/ceph-csi-rbd/README.md]: default is true in helm charts

Comment (pkalever) [reply charts/ceph-csi-rbd/README.md]: @Madhu-1 this is addressed now, thanks!

**Review by Madhu-1 — APPROVED** [review]

**Comment by Madhu-1** [comment]
Removed backport label please feel free to add it back if its required.

File: charts/ceph-csi-rbd/README.md
Comment (Rakshith-R) [reply charts/ceph-csi-rbd/README.md]: it should contain internal default that is false ? 

	flag.BoolVar(&conf.SetMetadata, "setmetadata", false, "set metadata on the volume")

Comment (pkalever) [reply charts/ceph-csi-rbd/README.md]: Hmm, right!

Comment (pkalever) [reply charts/ceph-csi-rbd/README.md]: Sorry got confused, I think its values from charts!

Comment (Madhu-1) [reply charts/ceph-csi-rbd/README.md]: AFAIK this is the documentation of the default values in values.yaml.

Comment (Rakshith-R) [reply charts/ceph-csi-rbd/README.md]: I believe it should internal default value since this will be used when no value in set for a variable in values.yaml.

Comment (pkalever) [reply charts/ceph-csi-rbd/README.md]: Lets come to a conclusion, we need this PR merged ASAP.

**Review by nixpanic — APPROVED** [review]

**Comment by nixpanic** [comment]
> Removed backport label please feel free to add it back if its required.

This is a new feature,  so it should not get backported.

**Review by yati1998 — APPROVED** [review]

**Comment by nixpanic** [comment]
This can be rebased with Mergify once #3216 is merged (that one is 1st in the queue atm).

**Comment by pkalever** [comment]
@Mergifyio rebase

**Comment by nixpanic** [comment]
> /retest ci/centos/mini-e2e-helm/k8s-1.22

```
Jun 28 15:51:43.874: FAIL: backend images not matching kubernetes resource count,image count 12 kubernetes resource count 11
```
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.22/detail/mini-e2e-helm_k8s-1.22/3648/pipeline)

**Comment by nixpanic** [comment]
> /retest ci/centos/k8s-e2e-external-storage/1.23

```
[Fail] External Storage [Driver: cephfs.csi.ceph.com] [Testpattern: Dynamic PV (default fs)] subPath [It] should support existing directories when readOnly specified in the volumeSource 
```
[logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/k8s-e2e-external-storage-1.23/detail/k8s-e2e-external-storage-1.23/2050/pipeline)

**Pull request merged**
